zallet/src/components/json_rpc/server/http_request_compatibility.rs

```diff
@@ -0,0 +1,292 @@
+//! Compatibility fixes for JSON-RPC HTTP requests.
+//!
+//! These fixes are applied at the HTTP level, before the RPC request is parsed.
+
+use std::future::Future;
+use std::pin::Pin;
+
+use futures::FutureExt;
+use http_body_util::BodyExt;
+use hyper::header;
+use jsonrpsee::{
+    core::BoxError,
+    server::{HttpBody, HttpRequest, HttpResponse},
+};
+use serde::{Deserialize, Serialize};
+use tower::Service;
+
+/// HTTP [`HttpRequestMiddleware`] with compatibility workarounds.
+///
+/// This middleware makes the following changes to HTTP requests:
+///
+/// ### Map between the client's JSON-RPC version and JSON-RPC 2.0.
+///
+/// [`jsonrpsee`] only supports JSON-RPC 2.0, while the existing Zcash ecosystem is used
+/// to communicating with `zcashd`'s "Bitcoin JSON-RPC" (a mix of 1.0, 1.1, and 2.0).
+///
+/// ### Add missing `content-type` HTTP header
+///
+/// Some RPC clients don't include a `content-type` HTTP header. But unlike web browsers,
+/// [`jsonrpsee`] does not do content sniffing.
+///
+/// If there is no `content-type` header, we assume the content is JSON, and let the
+/// parser error if we are incorrect.
+///
+/// ## Security
+///
+/// Any user-specified data in RPC requests is hex or base58check encoded. We assume the
+/// client validates data encodings before sending it on to Zallet. So any fixes Zallet
+/// performs won't change user-specified data.
+#[derive(Clone, Debug)]
+pub struct HttpRequestMiddleware<S> {
+    service: S,
+}
+
+impl<S> HttpRequestMiddleware<S> {
+    /// Create a new `HttpRequestMiddleware` with the given service.
+    pub fn new(service: S) -> Self {
+        Self { service }
+    }
+
+    /// Insert or replace client supplied `content-type` HTTP header to `application/json` in the following cases:
+    ///
+    /// - no `content-type` supplied.
+    /// - supplied `content-type` start with `text/plain`, for example:
+    ///   - `text/plain`
+    ///   - `text/plain;`
+    ///   - `text/plain; charset=utf-8`
+    ///
+    /// `application/json` is the only `content-type` accepted by the Zallet rpc endpoint:
+    ///
+    /// <https://github.com/paritytech/jsonrpc/blob/38af3c9439aa75481805edf6c05c6622a5ab1e70/http/src/handler.rs#L582-L584>
+    ///
+    /// # Security
+    ///
+    /// - `content-type` headers exist so that applications know they are speaking the correct protocol with the correct format.
+    ///   We can be a bit flexible, but there are some types (such as binary) we shouldn't allow.
+    ///   In particular, the "application/x-www-form-urlencoded" header should be rejected, so browser forms can't be used to attack
```

Comment:
I don't see where this is checked.

Reply:
This comment was written by the Zebra devs, and I think it might be out of date after their recent migration to the `jsonrpsee` crate. As noted in an earlier comment, the `jsonrpc` crate they previously used only accepts `application/json`, and it's probably the same for `jsonrpsee`.

Reply:
Indeed, `jsonrpsee` does the same thing: https://github.com/paritytech/jsonrpsee/blob/656f8bb0793c8e992d20b47c3d17e7a6c396fb8b/server/src/transport/http.rs#L14-L29